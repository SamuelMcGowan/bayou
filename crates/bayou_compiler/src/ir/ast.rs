use bayou_diagnostic::span::Span;

use super::ir::Type;
use super::{BinOp, Ident, Node, UnOp};

#[derive(Node!)]
pub struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub enum Item {
    FuncDecl(FuncDecl),
    ParseError,
}

#[derive(Node!)]
pub struct FuncDecl {
    pub name: Ident,
    pub ret_ty: Type,
    pub statements: Vec<Stmt>,
}

#[derive(Node!)]
pub enum Stmt {
    Assign { ident: Ident, expr: Expr },

    Return(Expr),

    ParseError,
}

#[derive(Node!)]
pub struct Expr {
    pub kind: ExprKind,
    pub span: Span,
}

impl Expr {
    pub fn new(kind: ExprKind, span: Span) -> Self {
        Self { kind, span }
    }
}

#[derive(Node!)]
pub enum ExprKind {
    Constant(i64),

    Var(Ident),

    UnOp {
        op: UnOp,
        expr: Box<Expr>,
    },

    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },

    Void,

    ParseError,
}
