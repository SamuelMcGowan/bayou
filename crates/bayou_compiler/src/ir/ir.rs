use super::{BinOp, Ident, Node, NodeCopy, UnOp};
use crate::symbols::LocalId;

pub struct FuncDecl {
    pub name: Ident,
    pub statements: Vec<Stmt>,
}

#[derive(Node!)]
pub enum Stmt {
    Assign { local: LocalId, expr: Expr },
    Return(Expr),
}

#[derive(Node!)]
pub struct Expr {
    pub kind: ExprKind,
    pub ty: Type,
}

#[derive(Node!)]
pub enum ExprKind {
    Constant(Constant),
    Var(LocalId),
    UnOp {
        op: UnOp,
        expr: Box<Expr>,
    },
    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },
}

#[derive(NodeCopy!)]
pub enum Constant {
    I64(i64),
}

impl Constant {
    pub fn ty(&self) -> Type {
        match self {
            Self::I64(_) => Type::I64,
        }
    }
}

#[derive(NodeCopy!)]
pub enum Type {
    I64,
}
