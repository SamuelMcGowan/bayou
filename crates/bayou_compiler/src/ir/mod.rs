use bayou_diagnostic::span::Span;

pub mod ast;
#[allow(clippy::module_inception)]
pub mod ir;
pub mod token;

pub type InternedStr = lasso::Spur;
pub type Interner = lasso::Rodeo;

derive_alias! {
    #[derive(Node!)] = #[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)];
    #[derive(NodeCopy!)] = #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize)];
}

#[derive(Node!, Copy)]
pub struct Ident {
    pub ident: InternedStr,
    pub span: Span,
}

#[derive(NodeCopy!)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,

    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    // Eq,
    // NotEq,

    // Gt,
    // Lt,
    // GtEq,
    // LtEq,
}

#[derive(NodeCopy!)]
pub enum UnOp {
    Negate,
    BitwiseInvert,
}

#[derive(NodeCopy!)]
pub struct Spanned<T> {
    pub node: T,
    pub span: Span,
}

impl<T> Spanned<T> {
    pub fn new(node: T, span: Span) -> Self {
        Self { node, span }
    }
}

impl<T, E> Spanned<Result<T, E>> {
    pub fn transpose(self) -> Result<Spanned<T>, E> {
        self.node.map(|node| Spanned {
            node,
            span: self.span,
        })
    }
}
