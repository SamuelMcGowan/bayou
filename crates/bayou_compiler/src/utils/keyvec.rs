use std::fmt;
use std::marker::PhantomData;
use std::ops::{Deref, Index, IndexMut};

pub struct KeyVec<K, V> {
    inner: Vec<V>,
    _phantom: PhantomData<*const K>,
}

impl<K: Key, V> KeyVec<K, V> {
    pub fn new() -> Self {
        Self::default()
    }

    #[must_use]
    pub fn insert(&mut self, value: V) -> K {
        let key = K::from_usize(self.inner.len());
        self.inner.push(value);
        key
    }

    pub fn get(&self, key: K) -> Option<&V> {
        self.inner.get(key.as_usize())
    }

    pub fn get_mut(&mut self, key: K) -> Option<&mut V> {
        self.inner.get_mut(key.as_usize())
    }
}

impl<K, V> Default for KeyVec<K, V> {
    fn default() -> Self {
        Self {
            inner: vec![],
            _phantom: PhantomData,
        }
    }
}

impl<K, V> Deref for KeyVec<K, V> {
    type Target = [V];

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<K: Key, V> Index<K> for KeyVec<K, V> {
    type Output = V;

    fn index(&self, key: K) -> &Self::Output {
        self.get(key).expect("key not found")
    }
}

impl<K: Key, V> IndexMut<K> for KeyVec<K, V> {
    fn index_mut(&mut self, key: K) -> &mut Self::Output {
        self.get_mut(key).expect("key not found")
    }
}

impl<K, V> IntoIterator for KeyVec<K, V> {
    type Item = V;

    type IntoIter = std::vec::IntoIter<V>;

    fn into_iter(self) -> Self::IntoIter {
        self.inner.into_iter()
    }
}

impl<K: Key + fmt::Debug, V: fmt::Debug> fmt::Debug for KeyVec<K, V> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_map()
            .entries(
                self.inner
                    .iter()
                    .enumerate()
                    .map(|(i, v)| (K::from_usize(i), v)),
            )
            .finish()
    }
}

impl<K, V: Clone> Clone for KeyVec<K, V> {
    fn clone(&self) -> Self {
        Self {
            inner: self.inner.clone(),
            _phantom: PhantomData,
        }
    }
}

impl<K, V: PartialEq> PartialEq for KeyVec<K, V> {
    fn eq(&self, other: &Self) -> bool {
        self.inner == other.inner
    }
}

pub trait Key: Copy {
    fn from_usize(n: usize) -> Self;
    fn as_usize(&self) -> usize;
}

macro_rules! declare_key_type {
(
    $v:vis struct $i:ident;
) => {
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize)]
    $v struct $i(pub usize);

    impl $crate::utils::keyvec::Key for $i {
        fn from_usize(n: usize) -> Self {
            Self(n)
        }

        fn as_usize(&self) -> usize {
            self.0
        }
    }
};
}
pub(crate) use declare_key_type;
