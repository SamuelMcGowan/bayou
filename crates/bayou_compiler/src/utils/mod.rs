pub mod keyvec;
pub mod peek;

macro_rules! assert_yaml_snapshot_with_source {
    ($test_name:expr; $source:expr => $output:expr) => {{
        insta::with_settings!({
            info => &$source,
            omit_expression => true,
        }, {
            insta::assert_yaml_snapshot!($test_name, $output);
        })
    }};
}
pub(crate) use assert_yaml_snapshot_with_source;
