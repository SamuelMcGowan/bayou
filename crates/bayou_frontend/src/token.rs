use bayou_diagnostic::span::Span;
use bayou_interner::Istr;

use crate::NodeCopy;

#[derive(NodeCopy!)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(NodeCopy!)]
pub enum TokenKind {
    Keyword(Keyword),
    Identifier(Istr),
    Integer(i64),
    Bool(bool),

    LBrace,
    RBrace,
    LParen,
    RParen,

    Dot,
    Colon,
    Comma,
    Semicolon,
    Bang,
    Arrow,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Assign,
    // TODO: uncomment when assignment is added
    // AddEq,
    // SubEq,
    // MulEq,
    // DivEq,
    // ModEq,

    // TODO: uncomment when boolean operations are added
    // EqEq,
    // NotEq,

    // Gt,
    // Lt,
    // GtEq,
    // LtEq,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    BitwiseInvert,
}

#[derive(NodeCopy!)]
pub enum Keyword {
    Func,
    Return,

    Let,
    If,
    Then,
    Else,

    I64,
    Bool,
    Void,
}

impl TokenKind {
    pub fn token_name(&self) -> &'static str {
        match self {
            TokenKind::Keyword(kw) => match kw {
                Keyword::Func => "keyword `func`",
                Keyword::Return => "keyword `return`",
                Keyword::Let => "keyword `let`",
                Keyword::If => "keyword `if`",
                Keyword::Then => "keyword `then`",
                Keyword::Else => "keyword `else`",
                Keyword::I64 => "keyword `i64`",
                Keyword::Bool => "keyword `bool`",
                Keyword::Void => "keyword `void`",
            },
            TokenKind::Identifier(_) => "identifier",
            TokenKind::Integer(_) => "integer",
            TokenKind::Bool(_) => "boolean",
            TokenKind::LBrace => "`{`",
            TokenKind::RBrace => "`}`",
            TokenKind::LParen => "`(`",
            TokenKind::RParen => "`)`",
            TokenKind::Dot => "`.`",
            TokenKind::Colon => "`:`",
            TokenKind::Comma => "`,`",
            TokenKind::Semicolon => "`;`",
            TokenKind::Bang => "`!`",
            TokenKind::Arrow => "`->`",
            TokenKind::Add => "`+`",
            TokenKind::Sub => "`-`",
            TokenKind::Mul => "`*`",
            TokenKind::Div => "`/`",
            TokenKind::Mod => "`%`",
            TokenKind::Assign => "`=`",
            TokenKind::BitwiseAnd => "`&`",
            TokenKind::BitwiseOr => "`|`",
            TokenKind::BitwiseXor => "`^`",
            TokenKind::BitwiseInvert => "`~`",
        }
    }
}
