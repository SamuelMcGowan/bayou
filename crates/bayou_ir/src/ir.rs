use bayou_interner::Interner;
use bayou_session::diagnostics::span::Span;
use bayou_session::sourcemap::SourceId;

use super::{BinOp, NodeCopyTraits, NodeTraits, UnOp};
use crate::symbols::{FuncId, LocalId, Symbols};
use crate::Type;

#[derive(NodeTraits!)]
pub struct Module {
    pub items: Vec<Item>,
}

// TODO: revisit borrowing issues
/// Additional information about a module that can't be stored in the [`Module`] type due
/// to borrowing issues.
pub struct ModuleContext {
    pub source_id: SourceId,
    pub symbols: Symbols,
}

pub struct Package {
    pub name: String,

    pub items: Vec<Item>,
    pub symbols: Symbols,
    pub interner: Interner,
}

#[derive(NodeTraits!)]
pub enum Item {
    FuncDecl(FuncDecl),
}

#[derive(NodeTraits!)]
pub struct FuncDecl {
    pub id: FuncId,
    pub block: Block,
}

#[derive(NodeTraits!)]
pub enum Stmt {
    Assign { local: LocalId, expr: Expr },
    Drop(Expr),
    Return(Expr),
}

#[derive(NodeTraits!)]
pub struct Block {
    pub statements: Vec<Stmt>,
    pub final_expr: Expr,
}

#[derive(NodeTraits!)]
pub struct Expr {
    pub kind: ExprKind,
    pub span: Span,
    pub ty: Option<Type>,
}

#[derive(NodeTraits!)]
pub enum ExprKind {
    Constant(Constant),
    Var(LocalId),
    UnOp {
        op: UnOp,
        expr: Box<Expr>,
    },
    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },
    Block(Box<Block>),
    If {
        cond: Box<Expr>,
        then: Box<Expr>,
        else_: Option<Box<Expr>>,
    },
}

#[derive(NodeCopyTraits!)]
pub enum Constant {
    I64(i64),
    Bool(bool),
    Void,
}

impl Constant {
    pub fn ty(&self) -> Type {
        match self {
            Self::I64(_) => Type::I64,
            Self::Bool(_) => Type::Bool,
            Self::Void => Type::Void,
        }
    }
}
