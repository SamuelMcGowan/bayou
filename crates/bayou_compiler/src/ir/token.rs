use bayou_diagnostic::span::Span;

use super::NodeCopy;
use crate::ir::InternedStr;

#[derive(NodeCopy!)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(NodeCopy!)]
pub enum TokenKind {
    Keyword(Keyword),
    Identifier(InternedStr),
    Integer(i64),

    LBrace,
    RBrace,
    LParen,
    RParen,
    Semicolon,
    Bang,
    Arrow,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Assign,
    // TODO: uncomment when assignment is added
    // AddEq,
    // SubEq,
    // MulEq,
    // DivEq,
    // ModEq,

    // TODO: uncomment when boolean operations are added
    // EqEq,
    // NotEq,

    // Gt,
    // Lt,
    // GtEq,
    // LtEq,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    BitwiseInvert,
}

#[derive(NodeCopy!)]
pub enum Keyword {
    Func,
    Return,

    Let,

    Void,
    I64,
}

impl TokenKind {
    pub fn token_name(&self) -> &'static str {
        match self {
            TokenKind::Keyword(kw) => match kw {
                Keyword::Func => "keyword `func`",
                Keyword::Return => "keyword `return`",
                Keyword::Let => "keyword `let`",
                Keyword::I64 => "keyword `i64`",
                Keyword::Void => "keyword `void`",
            },
            TokenKind::Identifier(_) => "identifier",
            TokenKind::Integer(_) => "integer",
            TokenKind::LBrace => "`{`",
            TokenKind::RBrace => "`}`",
            TokenKind::LParen => "`(`",
            TokenKind::RParen => "`)`",
            TokenKind::Semicolon => "`;`",
            TokenKind::Bang => "`!`",
            TokenKind::Arrow => "`->`",
            TokenKind::Add => "`+`",
            TokenKind::Sub => "`-`",
            TokenKind::Mul => "`*`",
            TokenKind::Div => "`/`",
            TokenKind::Mod => "`%`",
            TokenKind::Assign => "`=`",
            TokenKind::BitwiseAnd => "`&`",
            TokenKind::BitwiseOr => "`|`",
            TokenKind::BitwiseXor => "`^`",
            TokenKind::BitwiseInvert => "`~`",
        }
    }
}
