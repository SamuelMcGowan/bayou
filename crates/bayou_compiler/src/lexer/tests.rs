use super::{Lexer, LexerError};
use crate::ir::token::Token;
use crate::ir::Interner;
use crate::utils::assert_yaml_snapshot_with_source;

fn lex(source: &str) -> (Vec<Token>, Vec<LexerError>) {
    let mut interner = Interner::new();

    let lexer = Lexer::new(source, &mut interner);
    let (tokens, lexer_errors) = lexer.lex();

    (tokens.collect(), lexer_errors)
}

macro_rules! assert_lex {
    ($source:expr) => {{
        let source = $source;
        assert_yaml_snapshot_with_source!(source => lex(source));
    }};
}

#[test]
fn integer() {
    assert_lex!("100");
}

#[test]
fn integer_overflow() {
    assert_lex!("100000000000000000000");
}

#[test]
fn newlines() {
    assert_lex!("\nfunc\nmain\n(\n)\n->\nint\n{\nreturn\n0\n;\n}");
}

#[test]
fn no_newlines() {
    assert_lex!("func main()->int{return 0;}");
}

#[test]
fn spaces() {
    assert_lex!("  func  main  (  )  ->  int  {  return  0  ;  }");
}
