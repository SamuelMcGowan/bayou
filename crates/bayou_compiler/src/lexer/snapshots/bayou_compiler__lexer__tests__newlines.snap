---
source: crates/bayou_compiler/src/lexer/tests.rs
info: "\nfunc\nmain\n(\n)\n->\ni64\n{\nreturn\n0\n;\n}"
---
- - kind:
      Keyword: Func
    span:
      start: 1
      end: 5
  - kind:
      Identifier: 1
    span:
      start: 6
      end: 10
  - kind: LParen
    span:
      start: 11
      end: 12
  - kind: RParen
    span:
      start: 13
      end: 14
  - kind: Arrow
    span:
      start: 15
      end: 17
  - kind:
      Keyword: I64
    span:
      start: 18
      end: 21
  - kind: LBrace
    span:
      start: 22
      end: 23
  - kind:
      Keyword: Return
    span:
      start: 24
      end: 30
  - kind:
      Integer: 0
    span:
      start: 31
      end: 32
  - kind: Semicolon
    span:
      start: 33
      end: 34
  - kind: RBrace
    span:
      start: 35
      end: 36
- []

